protocol-units/da-sequencer/node/src/storage/mod.rs

```diff
@@ -188,7 +191,84 @@ impl DaSequencerStorage for Storage {
 	}
 
 	fn produce_next_block(&self) -> Result<Option<SequencerBlock>, DaSequencerError> {
-		todo!()
+		let cf_pending = self.db.cf_handle(cf::PENDING_TRANSACTIONS).ok_or_else(|| {
+			DaSequencerError::StorageAccess("Missing colomn family: pending transactions".into())
+		})?;
+
+		// Collect and deserialize all pending transactions
+		let mut transactions: Vec<Transaction> = Vec::new();
+
+		let iter = self.db.iterator_cf(&cf_pending, rocksdb::IteratorMode::Start);
+		for item in iter {
+			let (_key, value) = item.map_err(|e| DaSequencerError::RocksDbError(e.to_string()))?;
+			let tx: Transaction = bcs::from_bytes(&value)
+				.map_err(|e| DaSequencerError::Deserialization(e.to_string()))?;
+			transactions.push(tx);
+		}
+
+		if transactions.is_empty() {
+			return Ok(None);
+		}
+
+		// Sort transactions using their Ord implementation
+		transactions.sort();
+
+		let mut selected_txs = Vec::new();
+		let mut block_size: u64 = 0;
+
+		for tx in &transactions {
+			let tx_size = bcs::to_bytes(tx)
+				.map_err(|e| DaSequencerError::Deserialization(e.to_string()))?
+				.len() as u64; // The size of the tx is the len of the bytes Vec<u8>
+
+			if block_size + tx_size > MAX_SEQUENCER_BLOCK_SIZE {
+				break;
+			}
+
+			block_size += tx_size;
+			selected_txs.push(tx.clone());
+		}
+
+		// Notes that this method will never return less that 1
+		let height: u64 = self.determine_next_block_height()?.into();
+
+		let parent_height = BlockHeight(height - 1);
+		let parent_digest = self
+			.get_block_at_height(parent_height)?
+			.ok_or_else(|| DaSequencerError::StorageFormat("Missing parent block".into()))?
+			.get_block_digest();
+
+		// Build the block
+		let tx_set: BTreeSet<_> = selected_txs.clone().into_iter().collect();
+		let block = Block::new(BlockMetadata::default(), Id::new(parent_digest.0), tx_set);
+		let sequencer_block = SequencerBlock::try_new(self.determine_next_block_height()?, block)?;
```

Comment:
Use the calculated block height we use before. Don't recalculate.

Reply:
yah it's cause i'm calling into() there, to turn it into a u64, here it needs `Blockheight` I'll improve it so we only do the call once.

Reply:
BlockHeight is copy. Copy it before the into(). Something like ` (*&block_height).into() `
We should only call `determine_next_block_height() ` more than necessary because it queries the db.

Reply:
It's now only called once.